File: bms/go/go-gt/heap_pop/heap_pop.go
package test

var ()

const ()

func (h *Heap) pop() (i int64) {
	i = h.I[0]
	h.N--
	h.swap(0, h.N)
	h.down(0)
	return i
}

func (h Heap) swap(a, b int64) {
	i, j := h.I[a], h.I[b]
	h.I[a], h.I[b] = h.I[b], h.I[a]
	h.A[i], h.A[j] = b, a
}

func (h Heap) down(i int64) {
	for {
		left := 2*i + 1
		if left >= h.N {
			break
		}
		j := left
		if right := left + 1; right < h.N && !h.less(left, right) {
			j = right
		}
		if h.less(i, j) {
			break
		}
		h.swap(i, j)
		i = j
	}
}

func (h Heap) less(a, b int64) bool {
	i, j := h.I[a], h.I[b]
	return h.W[i] < h.W[j]
}

type Heap struct {
	N int64
	I []int64
	A []int64
	W []int64
}

File: translations_only/success/go/go-gt/heap_pop/gemini/CAPR/heap_pop.rs

use std::cmp::Ordering;

pub struct Heap {
    pub n: i64,
    pub i: Vec<i64>,
    pub a: Vec<i64>,
    pub w: Vec<i64>,
}

impl Heap {
    pub fn pop(&mut self) -> i64 {
        let i = self.i[0];
        self.n -= 1;
        self.swap(0, self.n as usize);
        self.down(0);
        i
    }

    pub fn swap(&mut self, a: usize, b: usize) {
        let i = self.i[a];
        let j = self.i[b];
        self.i[a] = self.i[b];
        self.i[b] = i;
        self.a[i as usize] = a as i64;
        self.a[j as usize] = b as i64;
    }

    pub fn down(&mut self, mut i: usize) {
        loop {
            let left = 2 * i + 1;
            if left >= self.n as usize {
                break;
            }
            let mut j = left;
            let right = left + 1;
            if right < self.n as usize && self.less(right, left) == Ordering::Less {
                j = right;
            }
            if self.less(i, j) == Ordering::Greater {
                break;
            }
            self.swap(i, j);
            i = j;
        }
    }

    pub fn less(&self, a: usize, b: usize) -> Ordering {
        self.w[self.i[a] as usize].cmp(&self.w[self.i[b] as usize])
    }
}

